src/linear_algebra/tensor_algebra.lean

```diff
@@ -0,0 +1,282 @@
+  commutes' := λ r x,
+  begin
+    dsimp only [],
+    rcases quot.exists_rep x with ⟨x,rfl⟩,
+    apply quot.sound,
+    apply rel.central_scalar,
+  end,
+  smul_def' := λ r x, by refl }
+
+/--
+The universal linear map `M →ₗ[R] tensor_algebra R M`.
+-/
+def univ : M →ₗ[R] (tensor_algebra R M) :=
+{ to_fun := λ m, quot.mk _ m,
+  map_add' := λ x y, by apply quot.sound; apply rel.add_lin,
+  map_smul' := λ r x, by apply quot.sound; apply rel.smul_lin }
+
+/--
+Given a linear map `f : M → A` where `A` is an `R`-algebra,
+`lift R M f` is the unique lift of `f` to a morphism of `R`-algebras
+`tensor_algebra R M → A`.
+-/
+def lift {A : Type u3} [semiring A] [algebra R A] (f : M →ₗ[R] A) : tensor_algebra R M →ₐ[R] A :=
+{ to_fun := λ a, quot.lift_on a (lift_fun _ _ f) $ λ a b h,
+  begin
+    induction h,
+    { change f _ = f _ + f _, simp, },
+    { change f _ = (algebra_map _ _ _) * f _, rw linear_map.map_smul,
+      exact algebra.smul_def h_r (f h_a) },
+    { change algebra_map _ _ _ = algebra_map _ _ _ + algebra_map _ _ _,
+      exact (algebra_map R A).map_add h_r h_s },
+    { change algebra_map _ _ _ = algebra_map _ _ _ * algebra_map _ _ _,
+      exact (algebra_map R A).map_mul h_r h_s },
+    { let G := lift_fun R M f,
+      change (algebra_map _ _ _) * G _ = G _,
+      exact algebra.commutes h_r (G h_a) },
+    { change _ + _ + _ = _ + (_ + _), rw add_assoc },
+    { change _ + _ = _ + _, rw add_comm, },
+    { let G := lift_fun R M f,
+      change (algebra_map _ _ _) + G _ = G _, simp, },
+    { change _ * _ * _ = _ * (_ * _), rw mul_assoc },
+    { let G := lift_fun R M f,
+      change (algebra_map _ _ _) * G _ = G _, simp, },
+    { let G := lift_fun R M f,
+      change G _ * (algebra_map _ _ _)= G _, simp, },
+    { change _ * (_ + _) = _ * _ + _ * _, rw left_distrib, },
+    { change (_ + _) * _ = _ * _ + _ * _, rw right_distrib, },
+    repeat { set G := lift_fun R M f,
+      change G _ + G _ = G _ + G _, rw h_ih, },
+    repeat { set G := lift_fun R M f,
+      change G _ * G _ = G _ * G _, rw h_ih, },
+    { change (algebra_map _ _ _) * _ = algebra_map _ _ _, simp },
+    { change _ * (algebra_map _ _ _) = algebra_map _ _ _, simp },
+  end,
+  map_one' := by change algebra_map _ _ _ = _; simp,
+  map_mul' :=
+  begin
+    intros x y,
+    rcases quot.exists_rep x with ⟨x,rfl⟩,
+    rcases quot.exists_rep y with ⟨y,rfl⟩,
+    refl,
+  end,
+  map_zero' := by change algebra_map _ _ _ = _; simp,
+  map_add' :=
+  begin
+    intros x y,
+    rcases quot.exists_rep x with ⟨x,rfl⟩,
+    rcases quot.exists_rep y with ⟨y,rfl⟩,
+    refl,
+  end,
+  commutes' := by tauto }
+
+variables {R M}
+
+theorem univ_comp_lift {A : Type u3} [semiring A] [algebra R A] (f : M →ₗ[R] A) :
+  (lift R M f).to_linear_map.comp (univ R M) = f := by {ext, refl}
+
+theorem lift_unique {A : Type u3} [semiring A] [algebra R A] (f : M →ₗ[R] A)
+  (g : tensor_algebra R M →ₐ[R] A) : g.to_linear_map.comp (univ R M) = f → g = lift R M f :=
```

Comment:
Would it make sense to combine these into a single `iff`?
```suggestion
theorem lift_unique {A : Type u3} [semiring A] [algebra R A] (f : M →ₗ[R] A)
  (g : tensor_algebra R M →ₐ[R] A) : g.to_linear_map.comp (univ R M) = f ↔ g = lift R M f :=
```

Reply:
Mathematically, yes, these two theorems are just a single logical equivalence. But my guess is that in practice both theorems are useful as they stand.

On a related note: should `univ_comp_lift` be tagged with `simp`?

Reply:
Yes, it should be a simp lemma.

Reply:
One possible advantage of making `lift_unique` an `iff` is that it could then also be used in `simp` and `rw`. (And possibly even be marked with `@[simp]`.

Reply:
Yes that's a good point. But I still think it's worthwhile to keep `univ_comp_lift` around, since I can envision a situation where I would want simp to simplify a bunch of compositions like those which appear in that theorem.

Reply:
Done!